File: LEETCODE/206. 反转链表.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 画个图贼清晰
 */
//递归
public ListNode reverseList(ListNode head) {
    if (head == null || head.next == null) 
        return head;
    ListNode reversed = reverseList(head.next);
    head.next.next = head;
    head.next = null;
    return reversed;
}

//迭代
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode pre = null, cur = head;
        while(cur!=null){
            ListNode curNextTemp = cur.next;
            cur.next = pre;
            pre = cur;
            cur = curNextTemp;
        }
        return pre;
    }
}

File: LEETCODE/kotlin/206. 反转链表.kt
//迭代
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        var pre: ListNode? = null
        var cur: ListNode? = head
        while (cur!=null) {
            var tmp: ListNode? = cur.next
            cur.next = pre
            pre = cur
            cur = tmp
        }
        return pre;
    }
}
//递归
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        if(head==null || head.next==null)
            return head
        val cur: ListNode? = reverseList(head.next)
        head.next.next = head
        head.next = null
        return cur
    }
}